src/UglyToad.PdfPig.DocumentLayoutAnalysis/KdTree.cs

```diff
@@ -0,0 +1,230 @@
+            {
+                var point = pivotPointFunc(pivot);
+                var currentNearestNode = node;
+                var currentDistance = distance(node.Value, point);
+
+                KdTreeNode<T> newNode = null;
+                double? newDist = null;
+
+                var pointValue = node.Depth == 0 ? point.X : point.Y;
+
+                if (pointValue < node.L)
+                {
+                    // start left
+                    (newNode, newDist) = FindNearestNeighbours(node.LeftChild, pivot, pivotPointFunc, distance);
+
+                    if (newDist.HasValue && newDist <= currentDistance && !newNode.Element.Equals(pivot))
+                    {
+                        currentDistance = newDist.Value;
+                        currentNearestNode = newNode;
+                    }
+
+                    if (node.RightChild != null && pointValue + currentDistance >= node.L)
+                    {
+                        (newNode, newDist) = FindNearestNeighbours(node.RightChild, pivot, pivotPointFunc, distance);
+                    }
+                }
+                else
+                {
+                    // start right
+                    (newNode, newDist) = FindNearestNeighbours(node.RightChild, pivot, pivotPointFunc, distance);
+
+                    if (newDist.HasValue && newDist <= currentDistance && !newNode.Element.Equals(pivot))
+                    {
+                        currentDistance = newDist.Value;
+                        currentNearestNode = newNode;
+                    }
+
+                    if (node.LeftChild != null && pointValue - currentDistance <= node.L)
+                    {
+                        (newNode, newDist) = FindNearestNeighbours(node.LeftChild, pivot, pivotPointFunc, distance);
+                    }
+                }
+
+                if (newDist.HasValue && newDist <= currentDistance && !newNode.Element.Equals(pivot))
+                {
+                    currentDistance = newDist.Value;
+                    currentNearestNode = newNode;
+                }
+
+                return (currentNearestNode, currentDistance);
+            }
+        }
+        #endregion
+
+        private class KdTreeLeaf<Q> : KdTreeNode<Q>
+        {
+            public override bool IsLeaf => true;
+
+            public KdTreeLeaf(PdfPoint l, Q element, int depth, int index)
+                : base(null, null, l, element, depth, index)
+            { }
+
+            public override string ToString()
+            {
+                return "Leaf->" + Value.ToString();
+            }
+        }
+
+        private class KdTreeNode<Q>
+        {
+            /// <summary>
+            /// Split value.
+            /// </summary>
+            public double L => Depth == 0 ? Value.X : Value.Y;
+
+            public PdfPoint Value { get; }
+
+            public KdTreeNode<Q> LeftChild { get; internal set; }
+
+            public KdTreeNode<Q> RightChild { get; internal set; }
```

Comment:
Can these setters be private / get only?

Reply:
Hi Eliot, we will need internal setters if we want to add an 'Add()' function to the tree. Good for me to leave it like this. Let me know

Reply:
Sure that's fine, I'll merge it in.